src/core/symbology/qgsrulebasedrenderer.cpp

```diff
@@ -1326,8 +1326,7 @@ QgsRuleBasedRenderer *QgsRuleBasedRenderer::convertFromRenderer( const QgsFeatur
     QString attr = categorizedRenderer->classAttribute();
     // categorizedAttr could be either an attribute name or an expression.
     // the only way to differentiate is to test it as an expression...
-    QgsExpression testExpr( attr );
-    if ( testExpr.hasParserError() || ( testExpr.isField() && !attr.startsWith( '\"' ) ) )
+    if ( QgsExpression::expressionToLayerFieldIndex( attr, layer ) != -1 && !attr.contains( '\"' ) )
```

Comment:
The test failures stem from the results when no layer is available. So I suggest using the new improved logic only when a layer is available, and falling back to the older method otherwise:

```suggestion
    bool isField = false;
    if ( layer )
    {
      isField = QgsExpression::expressionToLayerFieldIndex( attr, layer ) != -1;
    }
    else
    {
      QgsExpression testExpr( attr );
      isField = testExpr.hasParserError() || ( testExpr.isField() && !attr.startsWith( '\"' ) );
    }
    if ( isField )
```

Reply:
Ok that seems safe, though I'll keep the && condition when a layer is available, i've overlooked it but it seems that fields that are already quoted will be searched with the quote, I didn't dare to add quote handling logic in the core function to prevent that issue. But with only that things like "id" will not be found and will be turned into """id""" .

Reply:
Do you want to leave `QgsExpression::expressionToLayerFieldIndex` as is or add " support? I'm not sure if it's better to replace all the " upon parsing, only check at start and end if both are quotes or replace them if only two are counted. Or a mix of them to be safe, or do the quote-handlng separately.

Reply:
nvm, my error,

Reply:
Ok I'll escape quotes later,.